src/Bicep.Core/Workspaces/SourceFileGrouping.cs

```diff
@@ -21,23 +42,29 @@ public record UriResolutionResult(
         bool RequiresRestore,
         ErrorBuilderDelegate? ErrorBuilder);
 
-    public record ModuleSourceResolutionInfo(
-        IForeignTemplateReference ForeignTemplateReference,
-        ISourceFile ParentTemplateFile);
-
     public record SourceFileGrouping(
         IFileResolver FileResolver,
         Uri EntryFileUri,
         ImmutableDictionary<Uri, FileResolutionResult> FileResultByUri,
         ImmutableDictionary<ISourceFile, ImmutableDictionary<IForeignTemplateReference, UriResolutionResult>> UriResultByModule,
         ImmutableDictionary<ISourceFile, ImmutableHashSet<ISourceFile>> SourceFileParentLookup) : ISourceFileLookup
     {
-        public IEnumerable<ModuleSourceResolutionInfo> GetModulesToRestore()
+        public IEnumerable<IArtifactResolutionInfo> GetModulesToRestore()
             => UriResultByModule.SelectMany(
                 kvp => kvp.Value
                     .Where(entry => entry.Value.RequiresRestore)
                     .Select(entry => new ModuleSourceResolutionInfo(entry.Key, kvp.Key)));
 
+        public IEnumerable<IArtifactResolutionInfo> GetProvidersToRestore()
+        {
+            return SourceFiles.OfType<BicepSourceFile>()
+                .SelectMany(sourceFile => sourceFile.ProgramSyntax.Children.OfType<ProviderDeclarationSyntax>()
+                .Select(import => new ProviderSourceResolutionInfo(import, sourceFile)));
```

Comment:
These will get restored on every compilation (i.e., on each keypress in VS Code) since there's no check to make sure the referenced artifact requires restoration.

Reply:
I am not sure I follow, its being handled in what appears to me a similar way than `GetModulesToRestore` in the original implementation (see [here](https://github.com/Azure/bicep/blob/3319ec7e069073fe24dc2408438fca7185bce022/src/Bicep.Core/Workspaces/SourceFileGrouping.cs#L35)). I think that the check on wether a restore is needed happens downstream

Reply:
`GetModulesToRestore` [is checking each URI result to determine if a full restore is required](https://github.com/Azure/bicep/blob/3319ec7e069073fe24dc2408438fca7185bce022/src/Bicep.Core/Workspaces/SourceFileGrouping.cs#L38). There's no filter on restoration status [where that method is invoked in BicepCompiler](https://github.com/Azure/bicep/blob/3319ec7e069073fe24dc2408438fca7185bce022/src/Bicep.Core/BicepCompiler.cs#L52), and it looks like there is an early exit in [ModuleDispatcher::RestoreModules](https://github.com/Azure/bicep/blob/3319ec7e069073fe24dc2408438fca7185bce022/src/Bicep.Core/Registry/ModuleDispatcher.cs#L152-L156) that only applies if *all* the targets have already been restored.

This method is just returning a `ProviderSourceResolutionInfo` for every imported provider without filtering by restoration state. If you remove the `if (restorable is ProviderDeclarationSyntax) { continue; }` block in SourceFileGroupingBuilder, then `UriResultByModule` will also have restoration status information on provider artifacts.